### PR TITLE
Controller + Reconciler: Sink resolver + Lifecycle UTs

### PR DESCRIPTION
**Pull request by selim5**
Helps with https://github.com/googleinterns/knative-source-mongodb/issues/7 and https://github.com/googleinterns/knative-source-mongodb/issues/8

 Sink resolver + Lifecycle UTs

### CODE REVIEW

File: pkg/reconciler/mongodb/controller.go [hunk 1]
```diff
@@ -27,6 +27,8 @@ import (
 	configmap "knative.dev/pkg/configmap"
 	controller "knative.dev/pkg/controller"
 	logging "knative.dev/pkg/logging"
+	"knative.dev/pkg/resolver"
+
```
Comment (nachocano) [inline pkg/reconciler/mongodb/controller.go:31]: nit: no need for this extra line

File: pkg/apis/sources/v1alpha1/mongodbsource_lifecycle_test.go [hunk 2]
```diff
@@ -104,6 +104,7 @@ func TestMongoDbGetCondition(t *testing.T) {
 			m := &MongoDbSourceStatus{}
```
Comment (nachocano) [inline pkg/apis/sources/v1alpha1/mongodbsource_lifecycle_test.go:104]: can you also add one new UT, saying something like mark sink, deployed and connection established?

File: pkg/reconciler/mongodb/mongodbsource.go [hunk 3]
```diff
@@ -26,31 +28,42 @@ import (
 
 	v1alpha1 "github.com/googleinterns/knative-source-mongodb/pkg/apis/sources/v1alpha1"
 	mongodbsource "github.com/googleinterns/knative-source-mongodb/pkg/client/injection/reconciler/sources/v1alpha1/mongodbsource"
+	"knative.dev/pkg/apis"
+	duckv1 "knative.dev/pkg/apis/duck/v1"
+
```
Comment (nachocano) [inline pkg/reconciler/mongodb/mongodbsource.go:33]: nit: no need for this extra empty line

File: pkg/reconciler/mongodb/mongodbsource.go [hunk 4]
```diff
@@ -26,31 +28,42 @@ import (
...
 }
 
 // Reconciler implements controller.Reconciler for MongoDbSource resources.
 type Reconciler struct {
 	// Lister
 	secretLister corev1listers.SecretLister
+
+	sinkResolver *resolver.URIResolver
+
```
Comment (nachocano) [inline pkg/reconciler/mongodb/mongodbsource.go:58]: nit: no need for this extra empty line

File: pkg/reconciler/mongodb/mongodbsource.go [hunk 5]
```diff
@@ -63,6 +76,20 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, src *v1alpha1.MongoDbSou
...
+		src.Status.MarkNoSink("NotFound", "")
+		return err
+	}
+	src.Status.MarkSink(sinkURI)
+
+	// r.reconcileReceiveAdapter(ctx, src)
+	// TODO: make rsc.make receive adapter
+	// https://github.com/vaikas/postgressource/blob/b116b1097b87b9855a711f085f22996c522027bb/pkg/reconciler/deployment.go#L42
+	//  with lister, if doesnt exist create deployment
+	// if exist check hwo its different
```
Comment (nachocano) [inline pkg/reconciler/mongodb/mongodbsource.go:91]: nit: remove this

File: pkg/reconciler/mongodb/mongodbsource.go [hunk 6]
```diff
@@ -123,6 +150,29 @@ func (r *Reconciler) checkConnection(ctx context.Context, src *v1alpha1.MongoDbS
 	return nil
 }
 
+// checkSink checks the resolvability of the specified sink
+func (r *Reconciler) resolveSink(ctx context.Context, src *v1alpha1.MongoDbSource) (*apis.URL, reconciler.Event) {
```
Comment (nachocano) [inline pkg/reconciler/mongodb/mongodbsource.go:154]: you can just return (*apis.URL, error)

File: pkg/reconciler/mongodb/mongodbsource.go [hunk 7]
```diff
@@ -123,6 +150,29 @@ func (r *Reconciler) checkConnection(ctx context.Context, src *v1alpha1.MongoDbS
...
+	dest := src.Spec.Sink.DeepCopy()
+	if dest.Ref != nil {
+		if dest.Ref.Namespace == "" {
+			dest.Ref.Namespace = src.Namespace
+		}
+	}
+
+	sinkURI, err := r.sinkResolver.URIFromDestinationV1(*dest, src)
```
Comment (nachocano) [inline pkg/reconciler/mongodb/mongodbsource.go:162]: here just do `return r.sinkResolver.URIFromDestinationV1(*dest, src)`

File: pkg/reconciler/mongodb/mongodbsource.go [hunk 8]
```diff
@@ -123,6 +150,29 @@ func (r *Reconciler) checkConnection(ctx context.Context, src *v1alpha1.MongoDbS
...
+	if err != nil {
+		return nil, newWarningSinkNotFound(dest)
+	}
+
+	return sinkURI, nil
```
Comment (nachocano) [inline pkg/reconciler/mongodb/mongodbsource.go:167]: remove this part

File: pkg/reconciler/mongodb/mongodbsource.go [hunk 9]
```diff
@@ -123,6 +150,29 @@ func (r *Reconciler) checkConnection(ctx context.Context, src *v1alpha1.MongoDbS
...
+}
+
+// Helper Fucntion: Creates Event about Sink not being found.
+func newWarningSinkNotFound(sink *duckv1.Destination) reconciler.Event {
+	b, _ := json.Marshal(sink)
+	return reconciler.NewEvent(corev1.EventTypeWarning, "SinkNotFound", "Sink not found: %s", string(b))
+}
+
```
Comment (nachocano) [inline pkg/reconciler/mongodb/mongodbsource.go:175]: I think we can remove this

File: pkg/apis/sources/v1alpha1/mongodbsource_lifecycle_test.go (hunk 2 above)
Comment (selim5) [reply pkg/apis/sources/v1alpha1/mongodbsource_lifecycle_test.go:104]: Will do!

File: pkg/reconciler/mongodb/mongodbsource.go (hunk 4 above)
Comment (selim5) [reply pkg/reconciler/mongodb/mongodbsource.go:58]: I dont know why a lot of these extra lines dont exist on my local files??

File: pkg/apis/sources/v1alpha1/mongodbsource_lifecycle_test.go (hunk 2 above)
Comment (selim5) [reply pkg/apis/sources/v1alpha1/mongodbsource_lifecycle_test.go:104]: Oh forgot about that, will add tonite!

Comment (nachocano) [reply pkg/apis/sources/v1alpha1/mongodbsource_lifecycle_test.go:104]: ack, you can do it in a follow up

Comment (selim5) [reply pkg/apis/sources/v1alpha1/mongodbsource_lifecycle_test.go:104]: Will do!

**Pull request merged**